Keyboard shortcuts & command menu

hatemhosny opened the pull request:
<!--
  For Work In Progress Pull Requests, please use the Draft PR feature,
  see https://github.blog/2019-02-14-introducing-draft-pull-requests/ for further details.

  For a timely review/response, please avoid force-pushing additional
  commits if your PR already received reviews or comments.

  Before submitting a Pull Request, please ensure you've done the following:
  - 📖 Read the LiveCodes Contributing Guide: https://github.com/live-codes/livecodes/blob/HEAD/CONTRIBUTING.md.
  - 📖 Read the LiveCodes Code of Conduct: https://github.com/live-codes/livecodes/blob/HEAD/CODE_OF_CONDUCT.md.
  - 👷‍♀️ Create small PRs. In most cases, this will be possible.
  - ✅ Provide tests for your changes.
  - 📝 Use descriptive commit messages.
  - 📗 Update any related documentation and include any relevant screenshots.
  - 🌐 Use `window.deps.translateString` in .ts files and add `data-i18n` attributes in .html files to mark strings that needs to be translated.
-->

## What type of PR is this? (check all applicable)

- [x] ✨ Feature
- [ ] 🐛 Bug Fix
- [x] 📝 Documentation Update
- [ ] 🎨 Style
- [ ] ♻️ Code Refactor
- [ ] 🔥 Performance Improvements
- [ ] ✅ Test
- [ ] 🤖 Build
- [ ] 🔁 CI
- [ ] 📦 Chore (Release)
- [ ] ⏩ Revert
- [ ] 🌐 Internationalization / Translation

## Description

This PR adds 
- Command menu (Ctrl + K)
- Many keyboard shortcuts
- The UI shows the keyboard shortcuts (in title attributes or `<kbd>` elements)
- Keyboard shortcut menu showing the full collection of shortcuts

## Related Tickets & Documents

as discussed in #630 

## Mobile & Desktop Screenshots/Recordings

![image](https://github.com/user-attachments/assets/3478bb8c-c044-4262-a816-47e18f0e9e6a)

![image](https://github.com/user-attachments/assets/97b27c3f-2a6f-48c2-ba95-19e3c8065743)

![image](https://github.com/user-attachments/assets/57046a7d-bd2a-406c-89ed-aa73eebeaeb6)

![image](https://github.com/user-attachments/assets/136709d6-86ec-4358-8d3a-27266d602254)

![image](https://github.com/user-attachments/assets/98f92d92-af74-4805-8e02-a845590aa62a)

![image](https://github.com/user-attachments/assets/ca52502e-ccd7-4894-80dc-278eddede615)

![image](https://github.com/user-attachments/assets/5b345077-d447-4a23-80bc-c7d15895e6dc)


## Added tests?

- [ ] 👍 yes
- [ ] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help

## Added to documentations?

~~TODO: add to features section in docs~~

- [x] 📓 docs (./docs)
- [ ] 📕 storybook (./storybook)
- [ ] 📜 README.md
- [ ] 🙅 no documentation needed

## [optional] Are there any post-merge tasks we need to perform?

translations

@zyf722 @gigamaster 
Would you kindly have a look here at your convenience?
I'd like to know your general opinion, opinions about functionality, UI, selection of keyboard shortcuts and obviously if you find any issues.
Thank you.

zyf722 commented:
A very useful feature! I believe this will be a great addition to the app 🚀

Here are some issues and suggestions on it:

## Potential Issues
- Some actions like "About" and "Login" didn't work in some cases.
  - "Login" worked when I searched for "login" in the command palette, but not when no query was entered.
- Scrollbar styles for these two screens are not applied in my Google Chrome 118. I tested with the latest version of Edge, and it worked fine there, so I guess there might be some compatibility issues.

![scroll](https://github.com/user-attachments/assets/d420b5aa-d593-4010-a402-0ce432258b3e)

## Personal Suggestions
- From my point of view, the "Keyboard Shortcuts" screen might be a bit confusing due to its similar appearance to the command palette. I would personally prefer using a table- or grid-based layout.
  - From Google Docs
    ![image](https://github.com/user-attachments/assets/be70f61f-db79-4e68-bd9d-20be654e25a7)
  - From Figma
    ![image](https://github.com/user-attachments/assets/26f062ce-dd81-4b54-91c6-f11909274a16)
  - By the way, will users be able to customize the shortcuts in the future?

- From an i18n perspective, I suggest using a bilingual approach for the command palette, where the English text is displayed alongside the localized text. This way, users can search for commands in both their preferred language and English.
  - Take a reference from what we have in VSCode:
     ![](https://github.com/user-attachments/assets/9d3f07b2-9801-487a-8a9f-ba256dabafdb)
     ![](https://github.com/user-attachments/assets/445393d1-7b4e-41a5-bcec-396f95ec2b94)

hatemhosny commented:
Thank you, @zyf722 
These are all very useful comments

> * Some actions like "About" and "Login" didn't work in some cases.
>   
>   * "Login" worked when I searched for "login" in the command palette, but not when no query was entered.

I can reproduce that. I'll try to get a fix. Thank you.

> * Scrollbar styles for these two screens are not applied in my Google Chrome 118. I tested with the latest version of Edge, and it worked fine there, so I guess there might be some compatibility issues.

I did not get that in my browser, but I can think of a reason. I'll try to fix.

> * From my point of view, the "Keyboard Shortcuts" screen might be a bit confusing due to its similar appearance to the command palette. I would personally prefer using a table- or grid-based layout.
   
That's a good suggestion. It was just easy to implement, without code duplication by filtering the commands based if they have shortcuts and put it in a single list.
May be I can still get the data the same way but put that in a grid in a regular modal instead.
I'll explore and let you know.

>   * By the way, will users be able to customize the shortcuts in the future?

I did not think of that.
Probably not for now. Possible things to think about:
- We need to (locally) store the preferences as user data.
- Avoid conflicts with other app shortcuts and [editor shortcuts](https://code.visualstudio.com/docs/getstarted/keybindings#_basic-editing).
- Some shortcuts do not work in the browser (e.g. Ctrl + N), even if you prevent default.

I think I will leave it like this for now and revisit this suggestion later.

> * From an i18n perspective, I suggest using a bilingual approach for the command palette, where the English text is displayed alongside the localized text. This way, users can search for commands in both their preferred language and English.

Good Idea!
Do we need to show both? 
The library I use allows you to add a string of text to be used for search result without having to show it.

Thank you very much.
That was very helpful.

hatemhosny commented:
@zyf722 

> * Some actions like "About" and "Login" didn't work in some cases.
>   
>   * "Login" worked when I searched for "login" in the command palette, but not when no query was entered.

This was a race condition with double event firing. Fixed.

> * Scrollbar styles for these two screens are not applied in my Google Chrome 118. I tested with the latest version of Edge, and it worked fine there, so I guess there might be some compatibility issues.

This would be difficult to fix. The affected element is inside a web component with shadow DOM.
The styles are applied using `::part`, which is only available in [modern browsers](https://caniuse.com/mdn-css_selectors_part).
Otherwise we will have to change the library source code.

> * From my point of view, the "Keyboard Shortcuts" screen might be a bit confusing due to its similar appearance to the command palette. I would personally prefer using a table- or grid-based layout.

How about this? :)

![Screenshot (267)](https://github.com/user-attachments/assets/d2f2c9f4-0d8f-46b5-84eb-b64fed039bf2)

![image](https://github.com/user-attachments/assets/7e6d6735-06fa-4174-84e3-3614a9a4bff8)

> * From an i18n perspective, I suggest using a bilingual approach for the command palette, where the English text is displayed alongside the localized text. This way, users can search for commands in both their preferred language and English.

I added that. In languages other than English, English text will be displayed beneath each command. It can be searched.

![Screenshot 2025-01-09 131600](https://github.com/user-attachments/assets/95536b94-f857-4084-8c07-4e0718f3483b)

![Screenshot 2025-01-09 131833](https://github.com/user-attachments/assets/8c6817bc-8b6f-4caa-90ef-89580041a9fb)

![Screenshot 2025-01-09 132112](https://github.com/user-attachments/assets/cd9a401c-c36b-4560-a5de-f88889f74542)

![Screenshot 2025-01-09 132132](https://github.com/user-attachments/assets/958dcf0e-e04c-43ad-882a-d5fc7adc154d)

![Screenshot 2025-01-09 131939](https://github.com/user-attachments/assets/b890a765-cf98-4303-a856-56c7cd4abf62)


I think this is much better. Thank you for the suggestions.

zyf722 commented:
> Probably not for now. Possible things to think about:
> 
> * We need to (locally) store the preferences as user data.
> * Avoid conflicts with other app shortcuts and [editor shortcuts](https://code.visualstudio.com/docs/getstarted/keybindings#_basic-editing).
> * Some shortcuts do not work in the browser (e.g. Ctrl + N), even if you prevent default.

I agree with not implementing this feature for now. It indeed needs extra effort and consideration to implement this feature.

> This would be difficult to fix. The affected element is inside a web component with shadow DOM.
> The styles are applied using `::part`, which is only available in [modern browsers](https://caniuse.com/mdn-css_selectors_part).
> Otherwise we will have to change the library source code.

It's fully acceptable. Tried with BrowserStack and it does have something to do with the browser version. I believe we can leave it as is for now. Thank you for your efforts to fix this.

> How about this? :)

Looks quite nice! Does the tip above change with other editors when selected by user, like CodeJar and CodeMirror?

> I added that. In languages other than English, English text will be displayed beneath each command. It can be searched.

Great!

I think we could make it even more compact by removing the row gap and separator line between localized and English text. I believe the command palette will be more space-efficient this way.

hatemhosny commented:
.i18n-update-push

zyf722 commented:
It seems there might be another bug.

Some actions are not sorted as the first action in the list, even they match the user's query exactly:

![](https://github.com/user-attachments/assets/b1143a8b-6be1-421d-9c6e-939bfe824ec9)

![](https://github.com/user-attachments/assets/8c71c90b-1638-4949-85a3-6193722b6ed2)

![](https://github.com/user-attachments/assets/b2dc98d7-c5d8-4e80-9182-89c63611a9e3)

hatemhosny commented:
This is related to using fuzzy search. If we disable fuzzy search we get less results but better sorting.
I think I will do that.
Thank you.

![image](https://github.com/user-attachments/assets/d99115f9-9efb-4a30-9fcc-94557bc7cd32)

![image](https://github.com/user-attachments/assets/f9d3667a-d438-455d-b413-1e7e09bdeec2)

hatemhosny commented:
.i18n-update-pull

hatemhosny commented:
.i18n-update-pull

hatemhosny commented:
.i18n-update-pull

gigamaster commented:
This looks great. I'll try it asap. Well done ^_^/

gigamaster commented:
Translation ES, FR, IT, JA, PT

⚠ Note
It seems the same keyboard shortcut is used for sharing and saving !?

TODO - Test if any < tag- > is missing or overriten that could break the layout!

hatemhosny commented:
Welcome back @gigamaster . I hope you are doing very well and had good travel.

> Translation ES, FR, IT, JA, PT

Thank you very much. That's great.

> It seems the same keyboard shortcut is used for sharing and saving !?

- Save (Ctrl + S)
- Save as (Ctrl + Shift + S)
- Share (Ctrl + Alt + S)

Could you notice any inconsistent behaviour?

![image](https://github.com/user-attachments/assets/55fe7e49-aef1-419a-85bc-a57a664221e9)

> TODO - Test if any < tag- > is missing or overriten that could break the layout!

It seems they are good. When I filter by "Inconsistent placeholders", I get no results.
Am I correct @zyf722 ?
Also should we expect the scheduled sync to start a PR for the new changes on Lokalise master branch?

gigamaster commented:
Obviously I'm still jetlagged and must have misread the +alt and +shift keyboard shortcuts
After a nap I'll review my proxy settings as the SSH connection seems slow, or maybe it's just me 😅
I'll then merge and try the latest version. All good. Hope you're doing well ^_^/
